src/helper/helper.js

```diff
@@ -48,165 +48,163 @@ goog.module('helper');
  *
- * @constructor
- * @param {!Array.<!Object>} dataLayer The dataLayer to help with.
- * @param {function(!Object, !Object)=} opt_listener The callback function to
- *     execute when a new state gets pushed onto the dataLayer.
- * @param {boolean=} opt_listenToPast If true, the given listener will be
- *     executed for state changes that have already happened.
+ * You can retrieve values from the data model by using the helper's get method.
  */
-const DataLayerHelper = function(dataLayer, opt_listener, opt_listenToPast) {
-
+class DataLayerHelper {
   /**
-   * The dataLayer to help with.
-   * @type {!Array.<!Object>}
-   * @private
-   */
-  this.dataLayer_ = dataLayer;
-
-  /**
-   * The listener to notify of changes to the dataLayer.
-   * @type {function(!Object, !Object)}
-   * @private
+   * Creates a new helper object for the given dataLayer.
+   *
+   * @param {!Array.<!Object>} dataLayer The dataLayer to help with.
+   * @param {function(!Object, !Object)=} opt_listener The callback function to
+   *     execute when a new state gets pushed onto the dataLayer.
+   * @param {boolean=} opt_listenToPast If true, the given listener will be
+   *     executed for state changes that have already happened.
    */
-  this.listener_ = opt_listener || function() {};
-
-  /**
-   * The internal marker for checking if the listener is currently on the stack.
-   * @type {boolean}
-   * @private
-   */
-  this.executingListener_ = false;
+  constructor(dataLayer, opt_listener, opt_listenToPast) {
+    /**
+     * The dataLayer to help with.
+     * @type {!Array.<!Object>}
+     * @private
+     */
+    this.dataLayer_ = dataLayer;
+
+    /**
+     * The listener to notify of changes to the dataLayer.
+     * @type {function(!Object, !Object)}
+     * @private
+     */
+    this.listener_ = opt_listener || function() {};
+
+    /**
+     * The internal marker for checking if the listener is currently on the stack.
+     * @type {boolean}
+     * @private
+     */
+    this.executingListener_ = false;
+
+    /**
+     * The internal representation of the dataLayer's state at the time of the
+     * update currently being processed.
+     * @type {!Object}
+     * @private
+     */
+    this.model_ = {};
+
+    /**
+     * The internal queue of dataLayer updates that have not yet been processed.
+     * @type {Array.<Object>}
+     * @private
+     */
+    this.unprocessed_ = [];
+
+    /**
+     * The interface to the internal dataLayer model that is exposed to custom
+     * methods. Custom methods will the executed with this interface as the value
+     * of 'this', allowing users to manipulate the model using this.get and
+     * this.set.
+     * @type {!Object}
+     * @private
+     */
+    this.abstractModelInterface_ = buildAbstractModelInterface_(this);
+
+    // Process the existing/past states.
+    this.processStates_(dataLayer, !opt_listenToPast);
+
+    // Add listener for future state changes.
+    var oldPush = dataLayer.push;
```

Comment:
One more callout for a potential future change :)

Make sure to change these vars to consts/lets in a future PR if you're not planning on doing in this one now.

Reply:
Yes PR #37 has all the changes from vars to consts/lets, I would update this PR with those changes too but then I think I'll encounter some interesting merge conflicts.

Reply:
Ok, sounds good!